src/NuGet.Indexing/Extraction/DocumentCreator.cs

```diff
@@ -261,6 +262,57 @@ private static void AddRequiresLicenseAcceptance(Document document, IDictionary<
             }
         }
 
+        private static void AddPackageTypes(Document document, IDictionary<string, string> package)
+        {
+            string flattenedPackageTypes;
+            StringBuilder searchIndex = new StringBuilder();
+            if (package.TryGetValue("flattenedPackageTypes", out flattenedPackageTypes))
+            {
+                if (!string.IsNullOrWhiteSpace(flattenedPackageTypes))
+                {
+                    AddField(document, "FlattenedPackageTypes", flattenedPackageTypes, Field.Index.NOT_ANALYZED);
+
+                    using (var textWriter = new StringWriter())
+                    {
+                        using (var jsonWriter = new JsonTextWriter(textWriter))
+                        {
+                            jsonWriter.WriteStartArray();
+                            foreach (var packageType in flattenedPackageTypes.Split('|'))
+                            {
+                                string[] fields = packageType.Split(':');
+                                if (fields.Length > 0)
+                                {
+                                    jsonWriter.WriteStartObject();
+                                    jsonWriter.WritePropertyName("Type");
+                                    jsonWriter.WriteValue(fields[0]);
+                                    searchIndex.Append(fields[0]);
+                                    searchIndex.Append(" ");
```

Comment:
I think in general it would be a good idea to create an array list of the required fields for `searchIndex` and then convert it to a string delimited by <space> using `string.Join`. Personally I think that is a more readable format, no need to remove the added ending space in StringBuilder below.

To each his own, will leave it up to you.